### PR TITLE
[expo-gl-cpp] support reanimated worklets

### PR DESCRIPTION
**Pull request by wkozyra95**
# Why

First, step to support gl working in worklets #11267

### Limitations
- this implementation supports only worklets running UI thread(e.g. reanimted)
- this is a partial implementation (there will be few follow up PRs that are required for worklet support)
- required reanimated changes are not released yet https://github.com/software-mansion/react-native-reanimated/commit/c84229aa3f0da4f9eb3cbdba93243ed4b204b6f3

### Context

- https://github.com/expo/expo/pull/14023
- https://github.com/expo/expo/pull/14030
- https://github.com/expo/expo/pull/14033
- https://github.com/expo/expo/pull/14035
- https://github.com/expo/expo/pull/14036

# How

- if `global._WORKLET_RUNTIME` is set use the value as jsi::Runtime pointer and attach webgl renderer to main context and worklet context
- keep `JNIEnv` in thread-local variable, calls from c++ to java require it, but this object works only on the thread that it was passed from
- call RegisterThread() in UI thread before creating context,(this will allow triggering flush function from c++ for UI thread)

# Test Plan

I'm leaving testing for the last PR related to this feature, (it will be merged together before merging to master after review)

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-gl-cpp/cpp/TypedArrayApi.h [hunk 1]
```diff
@@ -60,7 +60,24 @@ struct typedArrayTypeMap<TypedArrayKind::Float64Array> {
   typedef double type;
 };
 
-void invalidateJsiPropNameIDCache();
+// instance of this class will invalidate PropNameIDCache when destructor is called
+// attach this object to global in specific jsi::Runtime to make sure lifecycle of
+// the cache object is connected to the lifecycle of the js runtime
```
Comment (tsapeta) [inline packages/expo-gl-cpp/cpp/TypedArrayApi.h:65]:
```suggestion
// Instance of this class will invalidate PropNameIDCache when destructor is called.
// Attach this object to global in specific jsi::Runtime to make sure lifecycle of
// the cache object is connected to the lifecycle of the JS runtime.
```

**Review by tsapeta — APPROVED** [review]

**Review by bbarthec — APPROVED** [review]
Looks fine to me 👍

**Review by Kudo — APPROVED** [review]
there are some format differences which might coming from clang-format version upgrade in `$ANDROID_NDK_HOME`. 
maybe we could use [this clang-format on npm](https://www.npmjs.com/package/clang-format) to pin down the clang-format version.

File: packages/expo-gl-cpp/cpp/TypedArrayApi.cpp [hunk 2]
```diff
@@ -45,14 +49,17 @@ class PropNameIDCache {
...
 };
 
 PropNameIDCache propNameIDCache;
 
-void invalidateJsiPropNameIDCache() {
+InvalidateCacheOnDestroy::InvalidateCacheOnDestroy(jsi::Runtime &runtime) {
+  key = reinterpret_cast<uintptr_t>(&runtime);
+}
+InvalidateCacheOnDestroy::~InvalidateCacheOnDestroy() {
   propNameIDCache.invalidate();
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/expo-gl-cpp/cpp/TypedArrayApi.cpp:63]: it should work, but it was not intentional, invalidate should take the key as an argument, just forgot about it, good catch

**Comment by wkozyra95** [comment]
> there are some format differences which might coming from clang-format version upgrade in $ANDROID_NDK_HOME.
maybe we could use this clang-format on npm to pin down the clang-format version.

yes formater behavior change significantly with a clang version, I was planning to pin the version inside the format script, but not as a part of this PR, I don't want to use that npm package because I think the entire toolchain should be pinned not only the formater

**Comment by Kudo** [comment]
> > there are some format differences which might coming from clang-format version upgrade in $ANDROID_NDK_HOME.
> > maybe we could use this clang-format on npm to pin down the clang-format version.
> 
> yes formater behavior change significantly with a clang version, I was planning to pin the version inside the format script, but not as a part of this PR, I don't want to use that npm package because I think the entire toolchain should be pinned not only the formater

that makes sense. thanks for your explanation.

**Pull request closed without merging**